### PR TITLE
HIVE-20967

### PR DESCRIPTION
**Pull request by ashutosh-bapat**
HIVE-20967

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java [hunk 1]
```diff
@@ -805,22 +811,89 @@ public Partition alterPartition(RawStore msdb, Warehouse wh, String catName, Str
...
+    if (conf.getBoolean(MetastoreConf.ConfVars.STRICT_MANAGED_TABLES.getHiveName(), false)) {
+      return;
+    }
+
+    // Do not allow changing location of a managed table as as alter event doesn't capture the
+    // new files list. So, it may cause data inconsistency.
+    boolean isChangingLocation = false;
+    if (ec.isSetProperties()) {
+      String alterType = ec.getProperties().get(ALTER_TABLE_OPERATION_TYPE);
+      if (alterType != null && alterType.equalsIgnoreCase(ALTERLOCATION)) {
```
Comment (maheshk114) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java]: add   "  && tbl.getTableType().equalsIgnoreCase(TableType.MANAGED_TABLE.name() . " and throw exception

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java [hunk 2]
```diff
@@ -805,22 +811,89 @@ public Partition alterPartition(RawStore msdb, Warehouse wh, String catName, Str
...
+                                                         EnvironmentContext ec)
+          throws InvalidOperationException {
+    // If the database is not replication source, nothing to do
+    if (!ReplChangeManager.isSourceOfReplication(db)) {
+      return;
+    }
+
+    // For now validate only the changes when strict managed tables is false. That's when there's
+    // scope for migration during replication, at least for now.
+    if (conf.getBoolean(MetastoreConf.ConfVars.STRICT_MANAGED_TABLES.getHiveName(), false)) {
```
Comment (maheshk114) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java]: this check can be removed

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java [hunk 3]
```diff
@@ -805,22 +811,89 @@ public Partition alterPartition(RawStore msdb, Warehouse wh, String catName, Str
...
-    // But, to be future proof, any table type conversion is restricted on a replication enabled DB.
-    if (!conf.getBoolean(MetastoreConf.ConfVars.STRICT_MANAGED_TABLES.getHiveName(), false)
-        && !oldTbl.getTableType().equalsIgnoreCase(newTbl.getTableType())
-        && ReplChangeManager.isSourceOfReplication(db)) {
+    // If the database is not replication source, nothing to do
+    if (!ReplChangeManager.isSourceOfReplication(db)) {
+      return;
+    }
+
+    // Do not allow changing location of a managed table as as alter event doesn't capture the
```
Comment (maheshk114) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java]: "table as as " ..two "as"

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java [hunk 4]
```diff
@@ -805,22 +811,89 @@ public Partition alterPartition(RawStore msdb, Warehouse wh, String catName, Str
...
+        String alterType = ec.getProperties().get(ALTER_TABLE_OPERATION_TYPE);
+        if (alterType != null && alterType.equalsIgnoreCase(ALTERLOCATION) &&
+            oldTbl.getTableType().equalsIgnoreCase(TableType.MANAGED_TABLE.name())) {
+          throw new InvalidOperationException("Cannot change location of a managed table " +
+                  TableName.getQualified(oldTbl.getCatName(),
+                          oldTbl.getDbName(), oldTbl.getTableName()) + " as it is enabled for replication.");
+        }
+    }
+
+    // Rest of the changes are need validation only when strict managed tables is false. That's
```
Comment (maheshk114) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java]: "changes are need" to "changes need"

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java (hunk 1 above)
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java]: Done.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java (hunk 2 above)
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java]: Done.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java (hunk 3 above)
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java]: Done. Thanks for catching it. Also removed it from table level validation.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java (hunk 4 above)
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java]: Done. Thanks for catching it.

**Pull request closed without merging**